kube-proxy handle duplicate endpoints

aojea opened the pull request:
Improve the EndpointSlice deduplication logic in kube-proxy's
EndpointSliceCache to ensure deterministic and correct handling of
duplicate endpoints (same IP:Port).

The cache now extracts the `endpoints.kubernetes.io/last-change-trigger-time`
annotation from EndpointSlices. When processing slices for a Service, they are
sorted by this timestamp. This ensures that endpoints from the most
recently updated Slice take precedence.

A `sequence` counter was added to track the order in which EndpointSlices
are updated in the cache to be used as a tie-breaker if trigger times
are equal or missing,preserving the order of arrival.

When duplicates exist (even within the same Slice), endpoints marked as `serving`
are preferred over non-serving ones.


/kind bug
Fixes #135266

```release-note
`kube-proxy` now correctly deduplicates endpoints when multiple EndpointSlices contain the same endpoint (IP:Port). It prioritizes the most recently updated EndpointSlice (based on `endpoints.kubernetes.io/last-change-trigger-time`) and prefers serving endpoints over non-serving ones.
```

pandaamanda commented:
/lgtm

danwinship commented:
Clarifying the above comment: 

- `CategorizeEndpoints` decides "which endpoint IPs kube-proxy should forward packets to"
- `addEndpoints` (which we are modifying here) tries to figure out "which `discoveryv1.Endpoint` corresponds to the Pod that is *actually* receiving traffic for `endpointIP`". The answer to that question might be an Endpoint that we want to forward traffic to, or it might be an Endpoint that we *don't* want to forward traffic to, but we're not deciding that here. We just need to pick the correct `Endpoint` here so that `CategorizeEndpoints` can make the correct decision later.

Consider two clusters: in both clusters, Service X is `externalTrafficPolicy: Local` and has three Endpoints:
- Endpoint A, with IP `10.128.0.2` which is local and ready
- Endpoint B, with IP `10.128.0.2` which is remote and serving-terminating
- Endpoint C, with IP `10.130.0.5` which is local and serving-terminating

However:
- In Cluster 1, Endpoint B is fake/wrong. Traffic sent to `10.128.0.2` will always be forwarded to Endpoint A.
- In Cluster 2, Endpoint A is fake/wrong. Traffic sent to `10.128.0.2` will always be forwarded to Endpoint B.

If `addEndpoints` prefers local endpoints to non-local endpoints, then it will keep Endpoint A, and discard Endpoint B. Then later, `CategorizeEndpoints` will be asked to consider Endpoint A and Endpoint C, and it would say "well, Endpoint A is local and ready, while Endpoint C is not ready, so we'll keep just Endpoint A" and so it would then send all traffic for Service X to `10.128.0.2`.

In cluster 1, that is the correct outcome. But in cluster 2, that's wrong: `10.128.0.2` *actually* routes to a remote Pod, violating the traffic policy. Kube-proxy *should* have picked Endpoint C instead in cluster 2, but it didn't realize that because `addEndpoints` had kept the wrong duplicate of `10.128.0.2`.

In other words, there is no reason for `addEndpoints` to look at `isLocal` (or `terminating`), because whether an Endpoint is "local" has nothing to do with whether the Endpoint is *real*, which is the thing `addEndpoints` actually needs to know.

And the only rule we've agreed on for deciding which Endpoint is real when there are duplicates is: if one Endpoint is `serving` and the others are not, then the `serving` one is the real one. So `addEndpoints` *should* consider `serving`, but it shouldn't consider the other conditions. (This rule won't help us pick the right Endpoint in the example above, but that's fine, because we have agreed that the example above is a broken cluster and kube-proxy is not expected to be able to figure out the right thing to do there.)

aojea committed:
kube-proxy: Improve EndpointSlice deduplication

Improve the EndpointSlice deduplication logic in kube-proxy's
EndpointSliceCache to ensure deterministic and correct handling of
duplicate endpoints (same IP:Port).

The cache now extracts the `endpoints.kubernetes.io/last-change-trigger-time`
annotation from EndpointSlices. When processing slices for a Service, they are
sorted by this timestamp. This ensures that endpoints from the most
recently updated Slice take precedence.

A `sequence` counter was added to track the order in which EndpointSlices
are updated in the cache to be used as a tie-breaker if trigger times
are equal or missing,preserving the order of arrival.

When duplicates exist (even within the same Slice), endpoints marked as `serving`
are preferred over non-serving ones.

danwinship commented on pkg/proxy/endpointslicecache_test.go:
```diff
@@ -311,6 +311,66 @@ func TestEndpointInfoByServicePort(t *testing.T) {
 				},
 			},
 		},
+		"two slices with duplicate IPs prefer last one to be processed (non local)": {
```
I don't think we should say that it actually prefers the last one.... We want to guarantee that we only keep one of them, but it doesn't matter which one.

Things I'd like better than having this test with this name:
- Split this test out into a new function so you can test "when there are duplicate Endpoints, we keep exactly 1" (but the test wouldn't care _which_ one, so that if the code changed later and started keeping the first one instead, the test would still pass).
- Keep the test as-is but rename it to "two serving slices with duplicate IPs, only one is kept", so that if the code changed later and started keeping the first one instead, it would be clearer that it's ok to change the test result to match the new code.
- Remove this test and only keep the "prefer serving" one below

aojea replied:
we prefer the last one because it will be the more recent received, we can overcomplicate by checking the object timestamps but I think this is just enough , if we have something in the map and we call AddEndpoints, we always overwrite the existing one, and also make the logic deterministic that simplify troubleshooting.

I think the existing patch and test cases cover the bugfix and backports, I prefer to leave for others the refactoring of the code

aojea commented:
@danwinship now I'm happier with the implementation, PTAL, added a more deterministic algorithm, as I do not think serving ALWAYS wins is correct, now the logic is:
1. more recent slice wins
2. latest processed slice within the cache wins (kube-proxy receives updates via a Watch stream, which guarantees ordering of events for a single resource)
3. serving wins

danwinship commented:
Hm... I was assuming that the existing code was doing the right thing for the "Endpoint gets moved between slices" case, and so we only had to worry about the "two pods have the same IP" case.

But I guess the existing code was slightly off even for the "Endpoint gets moved between slices" case, because it ignored the fact that the Endpoint from the more-recently-written slice would have a more-up-to-date set of `conditions` than the Endpoint from the less-recently-written slice. So if an Endpoint becomes non-`serving` at the same time as it gets moved to a different slice, then in that case we want to prefer the non-`serving` Endpoint in the new slice over the `serving` Endpoint in the old slice, because the non-`serving` one represents the current state.

But... you can't apply the "more-recent slice" rule to the "two pods with the same IP" case: if you have a "ghost" pod stuck in `Terminating` in slice 1, and a real `Ready` pod with the same IP in slice 2, you _always_ want to prefer the Endpoint for the real pod, regardless of which slice was written last.

So, to get both cases right, we have to distinguish whether we're looking at the "temporarily-duplicate Endpoints for a single pod" case or the "persistently-duplicate Endpoints for different pods" case.

So, if two `Endpoint`s have the same `addresses[0]`, then:
- If they have the same `targetRef`, then prefer the one from the more-recently-updated slice.
- If they have different `targetRef`s and only one of them is `serving`, then prefer the one that is `serving`
- Otherwise don't care / undefined.

----

The other thing about this patch is that it relies on kube-proxy perfectly observing all updates as they happen. If kube-proxy gets disconnected and has to resync (or if it starts up while the EndpointSlice controller is in the middle of rewriting a set of slices), it won't always know which slice is the newer one. This could be fixed if the EndpointSlice controller wrote out the sequence numbers to the actual EndpointSlices rather than kube-proxy just trying to compute them itself.

Alternatively, we _could_ just not bother with the "more-recently-updated slice" fix: if we pick the wrong Endpoint, it just means we keep using the older conditions until we receive the other updated slice, which is the same thing that would happen if we had just received the update slightly later. It's not the end of the world, and not even strictly "wrong".

/cc @robscott

danwinship commented on pkg/proxy/endpointslicecache.go:
```diff
@@ -168,7 +182,20 @@ func (cache *EndpointSliceCache) getEndpointsMap(serviceNN types.NamespacedName,
+		sliceDataList = append(sliceDataList, sliceData)
+	}
+	// Sort by trigger time and sequence
+	sort.Slice(sliceDataList, func(i, j int) bool {
```
While the existing code is not particularly optimized, it's worth considering that 99.9% of the time this sorting/copying will be unnecessary. It might be better to just put off the extra comparisions until we need them (by splitting out a `preferredEndpoint()` function or whatever that gets called below only when we realize we have a duplicate).

aojea commented:
> If kube-proxy gets disconnected and has to resync (or if it starts up while the EndpointSlice controller is in the middle of rewriting a set of slices), it won't always know which slice is the newer one. 

endpointSlices has an annotation with the timestamp of the last update.

Also, the watch will try to sync from the last known resource version IIRC ... but if not watch events will be always ordered, and you get the latest state of the object

> This could be fixed if the EndpointSlice controller wrote out the sequence numbers to the actual EndpointSlices rather than kube-proxy just trying to compute them itself.

There is already the resourceversion field that finally apimachinery is going to guarantee its comparable behavior for doing just what you say https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/5504-comparable-resource-version#helper-function

aojea commented:
> If they have the same targetRef, then prefer the one from the more-recently-updated slice.
> If they have different targetRefs and only one of them is serving, then prefer the one that is serving

I do not fully understand this logic and why we need to make assumptions on the targetReff, the controllers need to use the more recent data published, that is how the system is eventually consistent, if the data published is wrong then we should not workaround the problems on the consumers instead in the producers.

If there are same endpoint in different slices, we use the endpoint from the more up to date slice.
If the same slice has a duplicate endpoint, I picked serving based on your recommendation, but I still not complete following the logic

danwinship commented:
> if the data published is wrong then we should not workaround the problems on the consumers instead in the producers.

Right, but we agreed that the scenario in #135266 is _not_ wrong.

> the controllers need to use the more recent data published

We need to incorporate the new data into what we know, but in this case, the data from the more-recently-published slice doesn't change our interpretation of the overall situation. It's just saying "Hey, remember pod `caas-deployment-6ddfd47c58-w724s` that's been stuck in a Terminating state for a long time for some reason? Well, as of 23:12:54, it's _still_ stuck in a Terminating state". So nothing changed, and we still want to prefer the Endpoint for `caas-deployment-6ddfd47c58-wp7jk`.

(In the actual #135266 data, the two Endpoints are in the same slice, but you can see how it would all work if they were in different slices.)

aojea commented:
> Right, but we agreed that the scenario in #135266 is _not_ wrong.

I see, but it has the system that has to update the endpoint slices in order, and not us that make the assumption on the pod state ... in that case, the terminating state will be updated before the new Pod with the same IP is created
